lib/src/stream_transformers/buffer.dart

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+import 'dart:async';
+
+/// Creates a [StreamTransformer] which collects values and emits when it sees a
+/// value on [trigger].
+///
+/// If there are no pending values when [trigger] emits, the next value on the
+/// source Stream will immediately flow through. Otherwise, the pending values
+/// are released when [trigger] emits.
+///
+/// Errors from the source stream or the trigger are immediately forwarded to
+/// the output.
+StreamTransformer<T, List<T>> buffer<T>(Stream trigger) =>
+    new _Buffer(trigger, _collectToList);
+
+List<T> _collectToList<T>(T element, List<T> soFar) {
+  soFar ??= <T>[];
+  soFar.add(element);
+  return soFar;
+}
+
+/// A strategy for aggregating values.
+typedef R _Collect<T, R>(T element, R soFar);
+
+/// A StreamTransformer which aggregates values and emits when it sees a value
+/// on [_trigger].
+///
+/// If there are no pending values when [_trigger] emits the first value on the
+/// source Stream will immediately flow through. Otherwise, the pending values
+/// and released when [_trigger] emits.
+///
+/// Errors from the source stream or the trigger are immediately forwarded to
+/// the output.
+class _Buffer<T, R> implements StreamTransformer<T, R> {
+  final Stream _trigger;
+  final _Collect _collect;
```

Comment:
I'm also trying to decide if we want to expose a version of this which takes `collect` as an argument. It doesn't match what I think people would expect from `buffer` and I can't think of a better name... `aggregate`?

Originally in `build_runner` I had built it taking an argument so I could aggregate into a map instead of a list, but changed it to be a closer match to Rx version of buffer.

Do you think it would be worthwhile to expose it?